app/models/course.rb

```diff
@@ -148,4 +148,31 @@ def funding
       'fee'
     end
   end
+
+  def publish_sites
+    update_new_to_running
+    update_published_if_running
+    # as update_all is called and doesn't trigger a save this has to be manually tested
+    update(changed_at: Time.now.utc)
+  end
+
+  def publish_enrichment(current_user_id)
+    enrichments.where(status: 'draft')
+    .update_all(status: 'published',
+                updated_at: Time.now.utc,
+                last_published_timestamp_utc: Time.now.utc,
+                updated_by_user_id: current_user_id
+                )
```

Comment:
I don't think `update_all` is the right thing to use here, as it skips callbacks and validations (as you discovered with `changed_at`). I'm curious if there isn't a way to do this with a plain `update`, but it doesn't look like there is, so you'll probably have to iterate over each and run `update`.

Alternatively, as you point out there should only ever be one course enrichment in draft, you could just take the first one but I'd advise some way of guarding against there being multiples (raise an error if there is more than one course enrichment in draft).

Reply:
part 1 done
part 2 - not so sure about this at the moment as if there is an issue the current code would allow the user to continue unhindered, I'm starting to change my mind a bit on the fail early vs not breaking user journeys.

Reply:
@misaka what do you want to do about the guard thing if anything?

Reply:
@misaka @benilovj  I think in hindsight we modelled this wrong. Because draft and published are different things, they should (in a normalized database) be in different tables, then we could have applied proper database design and constraints to prevent bad data getting in. Worth bearing in mind as we look at reworking the structures later on. The draft could have been a field in the parent object, naturally enforcing the singular nature of them.